neqo-http3/src/hframe.rs

```diff
@@ -134,176 +134,161 @@ impl HFrameReader {
-                                | H3_FRAME_TYPE_MAX_PUSH_ID
-                                | H3_FRAME_TYPE_PUSH_PROMISE
-                                | H3_FRAME_TYPE_HEADERS => {
-                                    if len == 0 {
-                                        HFrameReaderState::Done
-                                    } else {
-                                        self.decoder = IncrementalDecoder::decode(
-                                            usize::try_from(len).or(Err(Error::HttpFrame))?,
-                                        );
-                                        HFrameReaderState::GetData
-                                    }
-                                }
-                                _ => {
-                                    if len == 0 {
-                                        self.decoder = IncrementalDecoder::decode_varint();
-                                        HFrameReaderState::BeforeFrame
-                                    } else {
-                                        self.decoder = IncrementalDecoder::ignore(
-                                            usize::try_from(len).or(Err(Error::HttpFrame))?,
-                                        );
-                                        HFrameReaderState::UnknownFrameDischargeData
-                                    }
-                                }
-                            };
-                        }
-                        IncrementalDecoderResult::InProgress => {}
-                        _ => panic!("We must be in one of the states above"),
-                    }
+            if fin {
+                if let HFrameReaderState::BeforeFrame { .. } = self.state {
+                    break Ok((None, fin));
+                } else {
+                    break Err(Error::HttpFrame);
                 }
-                HFrameReaderState::GetData => {
-                    match progress {
-                        IncrementalDecoderResult::Buffer(data) => {
-                            qtrace!(
-                                [conn],
-                                "received frame {}: {}",
-                                self.hframe_type,
-                                hex_with_len(&data[..])
-                            );
-                            self.payload = data;
-                            self.state = HFrameReaderState::Done;
-                        }
-                        IncrementalDecoderResult::InProgress => {}
-                        _ => panic!("We must be in one of the states above"),
+            }
+
+            if !read {
+                // There was no new data, exit the loop.
+                break Ok((None, false));
+            }
+        }
+    }
+
+    fn consume(&mut self, mut input: Decoder) -> Res<Option<HFrame>> {
+        match &mut self.state {
+            HFrameReaderState::BeforeFrame { decoder } | HFrameReaderState::GetType { decoder } => {
```

Comment:
Do you really need a separate state for "BeforeFrame"?

Reply:
It is used for one thing to decide if an end-of-stream is an error or not: in the middle of a frame it is an error otherwise it is not. I can indicate that in a different way as well.